rules/Transform/Rector/Class_/TestWithToDataProviderRector.php

```diff
@@ -0,0 +1,180 @@
+            return null;
+        }
+
+        $arrayItems = [];
+        foreach ($node->stmts as $classMethod) {
+            $hasChanged = false;
+            $phpDocInfo = $this->phpDocInfoFactory->createFromNode($classMethod);
+            if (! $phpDocInfo instanceof PhpDocInfo) {
+                return null;
+            }
+
+            $testWithPhpDocTagNodes = [
+                ...$phpDocInfo->getTagsByName('testWith'),
+                ...$phpDocInfo->getTagsByName('testwith'),
+            ];
+
+            if ($testWithPhpDocTagNodes === []) {
+                continue;
+            }
+
+            foreach ($testWithPhpDocTagNodes as $desiredTagValueNode) {
+                if (! $desiredTagValueNode->value instanceof GenericTagValueNode) {
+                    continue;
+                }
+
+                [$values] = $this->extractTestWithData($desiredTagValueNode->value);
+                $arrayItems[] = new Expr\ArrayItem($this->createArrayItem($values));
+
+                //cleanup
+                $hasChanged = $this->phpDocTagRemover->removeTagValueFromNode($phpDocInfo, $desiredTagValueNode);
+            }
+
+            if (!$hasChanged) {
+                continue;
+            }
+
+            $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($classMethod);
+            $phpDocInfo->addPhpDocTagNode(new PhpDocTagNode('@dataProvider', new GenericTagValueNode('dataProvider')));
+
+            $this->docBlockUpdater->updateRefactoredNodeWithPhpDocInfo($classMethod);
+        }
+
+        $array = new Expr\Array_($arrayItems);
+
+        $return = new Node\Stmt\Return_($array);
+        $providerMethod = new ClassMethod('dataProvider');
+        $providerMethod->flags = Class_::MODIFIER_PUBLIC;
+        $providerMethod->stmts[] = $return;
+        $this->classInsertManipulator->addAsFirstMethod($node, $providerMethod);
+
+        return $node;
+    }
+
+    private function extractTestWithData(GenericTagValueNode $genericTagValueNode): array
+    {
+        $testWithItems = explode("\n", trim($genericTagValueNode->value));
+        $jsonArray = [];
+
+        foreach ($testWithItems as $testWithItem) {
+            $jsonArray[] = Json::decode(trim($testWithItem), Json::FORCE_ARRAY);
+        }
+
+        return $jsonArray;
+    }
+
+    private function createArrayItem(array $data): Expr\Array_
+    {
+        $values = [];
+        foreach ($data as $item) {
+            $values[] = new ArrayItem($this->parseArrayItemValue($item));
+        }
+
+        return new Expr\Array_($values);
+    }
+
+    private function parseArrayItemValue(mixed $value): Expr
+    {
+        if(is_array($value)) {
+            return $this->createArrayItem($value);
+        }
+
+        if(is_string($value)) {
+            return new Node\Scalar\String_($value);
+        }
+
+        if(is_integer($value)) {
+            return new Node\Scalar\LNumber($value);
+        }
+
+        return new Node\Scalar\DNumber($value);
```

Comment:
Hello guys

I need some support here :(, I do not know if it is the best/only way to parse those values.

But to make this works I need to use `ArrayItem` and ArrayItem requires a object `Expr`.


But for some cases like `bool` and `null` values, I do not have a Scalar class for this (well I did not find).

it is just missing make those tests pass, you can check into `feature3.php.inc`.

cc @TomasVotruba or @samsonasik